File: kitchen-agent/src/api/providers.py
"""
api/providers.py
────────────────
Provider information and app metadata endpoints.

Routes:
  GET /api/providers          → list available providers
  GET /api/providers/active   → active provider and model
  GET /api/app-info           → app metadata
"""
from __future__ import annotations

from fastapi import APIRouter

from src.config import settings
from src.providers.config import (
    ANTHROPIC_DEFAULT_MODEL,
    GEMINI_DEFAULT_MODEL,
    MIMO_DEFAULT_MODEL,
)
from src.schemas import ActiveProvider, AppInfo, ModelInfo, ProviderInfo

router = APIRouter()


# ── Provider catalogue (static) ────────────────────────────────────

_PROVIDER_CATALOGUE: list[ProviderInfo] = [
    ProviderInfo(
        id="gemini",
        label="Google Gemini",
        default_model=GEMINI_DEFAULT_MODEL,
        models=[
            ModelInfo(id="gemini-3.1-pro-preview", label="Gemini 3.1 Pro", context_k=1000),
            ModelInfo(id="gemini-3.5-flash", label="Gemini 3.5 Flash", context_k=1000),
        ],
    ),
    ProviderInfo(
        id="anthropic",
        label="Anthropic Claude",
        default_model=ANTHROPIC_DEFAULT_MODEL,
        models=[
            ModelInfo(id="claude-opus-4-8", label="Claude Opus 4.8", context_k=200),
            ModelInfo(id="claude-sonnet-4-6", label="Claude Sonnet 4.6", context_k=200),
        ],
    ),
    ProviderInfo(
        id="mimo",
        label="Xiaomi MiMo",
        default_model=MIMO_DEFAULT_MODEL,
        models=[
            ModelInfo(id="mimo-v2.5-pro", label="MiMo V2.5 Pro", context_k=1000),
            ModelInfo(id="mimo-v2.5", label="MiMo V2.5", context_k=1000),
        ],
    ),
]

_PROVIDER_MAP: dict[str, ProviderInfo] = {p.id: p for p in _PROVIDER_CATALOGUE}


@router.get("/api/providers", response_model=list[ProviderInfo])
def list_providers() -> list[ProviderInfo]:
    """Returns the full provider + model catalogue."""
    return _PROVIDER_CATALOGUE


def get_default_model(provider_id: str) -> str:
    """Get the catalogue default model for a provider.  Single source of truth."""
    entry = _PROVIDER_MAP.get(provider_id)
    return entry.default_model if entry else ""


@router.get("/api/providers/active", response_model=ActiveProvider)
def get_active_provider() -> ActiveProvider:
    """Returns the server-configured default provider and model."""
    provider = settings.llm_provider
    # Settings fields allow env-var overrides; fall back to catalogue default.
    model = getattr(settings, f"{provider}_model", None) or get_default_model(provider)
    return ActiveProvider(provider=provider, model=model)


@router.get("/api/app-info", response_model=AppInfo)
def get_app_info() -> AppInfo:
    """Returns domain branding metadata."""
    return AppInfo(title=settings.app_title, description=settings.app_description)

File: kitchen-agent/frontend/src/lib/providers.ts
/**
 * src/lib/providers.ts
 * =====================
 * Static catalog of available LLM providers and their model lists.
 *
 * This is the v1 "hardcoded" approach from §7 of the provider integration guide.
 * Once the backend exposes GET /api/providers, this catalog will be replaced
 * by a live fetch — but it is used as the fallback and initial seed.
 *
 * To add a new model: extend the `models` array for the relevant provider.
 * To add a new provider: push a new entry to PROVIDERS.
 */

export type ModelInfo = {
  id: string;
  label: string;
  context_k: number; // context window in thousands of tokens
};

export type ProviderInfo = {
  id: string;
  label: string;
  default_model: string;
  models: ModelInfo[];
};

export const PROVIDERS: ProviderInfo[] = [
  {
    id: "gemini",
    label: "Google Gemini",
    default_model: "gemini-3.1-pro-preview",
    models: [
      { id: "gemini-3.1-pro-preview", label: "Gemini 3.1 Pro", context_k: 1000 },
      { id: "gemini-3.5-flash", label: "Gemini 3.5 Flash", context_k: 1000 },
    ],
  },
  {
    id: "anthropic",
    label: "Anthropic Claude",
    default_model: "claude-sonnet-4-6",
    models: [
      { id: "claude-opus-4-8", label: "Claude Opus 4.8", context_k: 200 },
      { id: "claude-sonnet-4-6", label: "Claude Sonnet 4.6", context_k: 200 },
    ],
  },
  {
    id: "mimo",
    label: "Xiaomi MiMo",
    default_model: "mimo-v2.5-pro",
    models: [
      { id: "mimo-v2.5-pro", label: "MiMo V2.5 Pro", context_k: 1000 },
      { id: "mimo-v2.5", label: "MiMo V2.5", context_k: 1000 },
    ],
  },
];

/** Look up a provider entry by id. Returns undefined when not found. */
export function findProvider(id: string): ProviderInfo | undefined {
  return PROVIDERS.find((p) => p.id === id);
}

/** Look up a model inside a specific provider. Returns undefined when not found. */
export function findModel(providerId: string, modelId: string): ModelInfo | undefined {
  return findProvider(providerId)?.models.find((m) => m.id === modelId);
}
